src/v/lsm/io/disk_persistence.cc

```diff
@@ -0,0 +1,352 @@
+        try {
+            co_await _file.close();
+        } catch (const std::system_error& err) {
+            throw io_error_exception(err.code(), "io error closing: {}", err);
+        } catch (...) {
+            throw io_error_exception(
+              "io error closing: {}", std::current_exception());
+        }
+    }
+
+    fmt::iterator format_to(fmt::iterator it) const override {
+        return fmt::format_to(it, "{{path={}}}", _path);
+    }
+
+private:
+    std::filesystem::path _path;
+    ss::file _file;
+};
+
+class disk_seq_file_writer : public sequential_file_writer {
+public:
+    disk_seq_file_writer(
+      std::filesystem::path path, ss::output_stream<char> stream)
+      : _path(std::move(path))
+      , _stream(std::move(stream)) {}
+
+    ss::future<> append(iobuf buf) override {
+        try {
+            for (auto& frag : buf) {
+                co_await _stream.write(frag.get(), frag.size());
+            }
+        } catch (const std::system_error& err) {
+            throw io_error_exception(err.code(), "io error writing: {}", err);
+        } catch (...) {
+            throw io_error_exception(
+              "io error writing: {}", std::current_exception());
+        }
+    }
+    ss::future<> append(ioarray array) override {
+        try {
+            for (const auto& frag : array.as_iovec()) {
+                co_await _stream.write(
+                  static_cast<const char*>(frag.iov_base), frag.iov_len);
+            }
+        } catch (const std::system_error& err) {
+            throw io_error_exception(err.code(), "io error writing: {}", err);
+        } catch (...) {
+            throw io_error_exception(
+              "io error writing: {}", std::current_exception());
+        }
+    }
+    ss::future<> close() override {
+        try {
+            co_await _stream.close();
+        } catch (const std::system_error& err) {
+            throw io_error_exception(err.code(), "io error closing: {}", err);
+        } catch (...) {
+            throw io_error_exception(
+              "io error closing: {}", std::current_exception());
+        }
+    }
+    fmt::iterator format_to(fmt::iterator it) const override {
+        return fmt::format_to(it, "{{path={}}}", _path);
+    }
+
+private:
+    std::filesystem::path _path;
+    ss::output_stream<char> _stream;
+};
+
+class impl : public persistence {
+public:
+    explicit impl(std::filesystem::path root)
+      : _root(std::move(root)) {}
+
+    ~impl() override = default;
+
+    ss::future<optional_pointer<sequential_file_reader>>
+    open_sequential_reader(std::string_view name) override {
+        try {
+            auto filepath = path(name);
+            auto file = ss::open_file_dma(
+              filepath.native(), ss::open_flags::ro);
+            auto stream = co_await ss::with_file_close_on_failure(
+              std::move(file), [](ss::file f) {
+                  return ss::make_file_input_stream(std::move(f));
```

Comment:
Maybe it makes sense to use `file_input_stream_options`. We have few config options like `storage_read_buffer_size` and `storage_read_readahead_count` in the config already. Number of readers is often much larger than number of writers so these parameters might prove to be useful to control memory utilization.